### PR TITLE
Added Pantheon Online Accounts documentation.

### PR DESCRIPTION
**Pull request by tintou**

### CODE REVIEW

**Comment by lewisgoddard** [comment]
Looks good web-wise, I have no idea about the actual document though.

File: docs/code/getting-started.md [hunk 1]
```diff
@@ -675,4 +675,141 @@ Let’s recap what we learned in this section:
...
+Most services allows you to only use “tokens” and thus not having to deal with the user password. Token are like passwords as they allow you to interact with the service (here add, modify and remove contacts) but unlike passwords, you are restricted to a specific area of the whole service (you can’t access Google Drive documents with a token made for Contacts) and in top of that, tokens are revocable. The Pantheon Online Account service is a central place to retrieve, store and refresh tokens when they are needed so you don’t have to re-implement all the Authentication procedure. All you need to know is to use the API of the provider of your choice.
+
+## Create a Provider {#poa-create-provider}
+
+There are several default providers shipped with elementary, and much more available in the AppCenter. If the provider you want is already available, jump to the next part as you *MUST* use it. Otherwise, here is how to create a new one.
```
Comment (danrabbit) [inline docs/code/getting-started.md]: Link on "jump to the next part"?

File: docs/code/getting-started.md [hunk 2]
```diff
@@ -675,4 +675,141 @@ Let’s recap what we learned in this section:
...
+* The `translations` is the domain that should be installed and provide translations for the description
+* The `plugin` is the name of the plugin POA will use to handle the account
+
+As you can see, the biggest amount of lines are contained in the `template` key. They represent everything POA will need to authenticate, the required keys can variate between each methods and mechanisms so you might want to take a look at the keys that are supported by them.
+Nevertheless, a common part of the `template` is that you always have to specify the `auth/method` and `auth/mechanisms` settings, then you have to put all the supported keys in the `auth/$method/$mechanism` group. If any key is missing, the backend will fail and return you an explicit error message in the terminal.
+
+## Create a Service {#poa-create-service}
+
+As in the Create a Provider section, there are several default services shipped with elementary, and much more available in the AppCenter. If the service you want is already available, jump to the next part as you *MUST* use it. Otherwise, here is how to create a new one.
```
Comment (danrabbit) [inline docs/code/getting-started.md]: Link "Create a Provider"

Comment (danrabbit) [inline docs/code/getting-started.md]: Link "jump to the next part"

File: docs/code/getting-started.md [hunk 3]
```diff
@@ -675,4 +675,141 @@ Let’s recap what we learned in this section:
...
+The Pantheon Online Accounts service has already the major authentication protocols built-in, it is possible to create yourself everything to support a new protocol but we won't cover this here.
+Each service has a corresponding authentication method and mechanism (if the authentication isn't standard, those are often equal), here is a non-exhaustive list of the most used ones:
+
+|   Methods  |                Mechanisms                |
+|  --------  | ---------------------------------------- |
+|  `oauth1`  | `PLAINTEXT` or `HMAC-SHA1` or `RSA-SHA1` |
+|  `oauth2`  |       `web_server` or `user_agent`       |
+| `password` |                `password`                |
+
+To create a provider, you have to ship a .provider file in "/usr/share/accounts/providers/", it is better to name the file as `%provider.provider` (here it’s `google.provider` for example)
```
Comment (danrabbit) [inline docs/code/getting-started.md]: I think this needs more explanation. Where to I ship this file with my app? In the "data" directory? How do I install it to /usr/share/accounts/providers? What do I need to do to my CMake or with packaging?

File: docs/code/getting-started.md [hunk 4]
```diff
@@ -675,4 +675,141 @@ Let’s recap what we learned in this section:
...
+* The `plugin` is the name of the plugin POA will use to handle the account
+
+As you can see, the biggest amount of lines are contained in the `template` key. They represent everything POA will need to authenticate, the required keys can variate between each methods and mechanisms so you might want to take a look at the keys that are supported by them.
+Nevertheless, a common part of the `template` is that you always have to specify the `auth/method` and `auth/mechanisms` settings, then you have to put all the supported keys in the `auth/$method/$mechanism` group. If any key is missing, the backend will fail and return you an explicit error message in the terminal.
+
+## Create a Service {#poa-create-service}
+
+As in the Create a Provider section, there are several default services shipped with elementary, and much more available in the AppCenter. If the service you want is already available, jump to the next part as you *MUST* use it. Otherwise, here is how to create a new one.
+The .service file should be stored in "/usr/share/accounts/services/", it is better to name the file as `%provider-%service.service` (here it’s `google-contacts.service` for example)
```
Comment (danrabbit) [inline docs/code/getting-started.md]: Same question as the above file. Where do I store this in my project? How do I install it with CMake? Do I have to do anything with packaging?

File: docs/code/getting-started.md [hunk 5]
```diff
@@ -675,4 +675,141 @@ Let’s recap what we learned in this section:
...
+                </service>
+            </services>
+            <service-types>
+                <service-type id="contacts">
+                </service-type>
+            </service-types>
+        </application>
+
+In some cases, your application will handle all kind of services (such as e-mails). If so, you can just omit the `services` and just fill the `service-types` you want.
+
```
Comment (danrabbit) [inline docs/code/getting-started.md]: This seems really incomplete. I've created these files, now what? How does my app actually interact with the data that these things provide?

**Comment by tintou** [comment]
Okay, I fixed the missing links and added the addition of the CMakeLists.txt
But you're right, I also need to merge the integration into an application
I already have a Google Document that I have to adapt in order to have the same amount of description as the rest of the documentation

File: docs/code/getting-started.md [hunk 6]
```diff
@@ -675,4 +675,162 @@ Let’s recap what we learned in this section:
...
+                    <setting name="mechanism">user_agent</setting>
+                    <group name="oauth">
+                        <group name="user_agent">
+                            <setting type="boolean" name="ForceClientAuthViaRequestBody">true</setting>
+                            <setting name="AuthHost">accounts.google.com</setting>
+                            <setting name="TokenHost">accounts.google.com</setting>
+                            <setting name="AuthPath">/o/oauth2/auth</setting>
+                            <setting name="TokenPath">/o/oauth2/token</setting>
+                            <setting type="int" name="UiPolicy">0</setting>
+                            <setting name="RedirectUri">http://elementaryos.org</setting>
```
Comment (nathandyer) [inline docs/code/getting-started.md]: Here the RedirectUri is listed as "elementaryos.org," yet above it is "elementary.io." I imagine that we want to be consistent. Also, it's probably worth mentioning to a new developer that the URI you choose isn't all that important, but that it MUST match the redirect URI you provide to the actual service provider (Twitter, Facebook, etc.) when you sign up to make an application (unless I'm wrong about that part).

**Comment by nathandyer** [comment]
What's there is pretty good. In case you all weren't aware/forgot, [here](https://docs.google.com/document/d/16l4pvhJdcL0V37JKojkyC3z8gGsFVLATDmoxPsSyR7Q/edit?usp=sharing) is a link to a Google Doc with documentation that someone has already written about how to retrieve the token from POA in a typical Vala app. It's even based around Google Contacts, so it should be easy to copy/paste from it.

**Comment by danrabbit** [comment]
Closing because this hasn't moved in over a year

**Pull request closed without merging**